[OptimistLabyrinth] 🚀 2단계 - 지하철 노선 기능

OptimistLabyrinth opened the pull request:
1단계 미션을 진행한 뒤로 계속 고민하던 부분인데요
인수테스트를 가독성을 증대할 수 있는 방향으로 리팩터링 했고
이후에 2단계 미션을 진행했습니다.

2단계 미션의 요구사항을 읽으며
지하철 노선(Line) 과 지하철 역(Station) 의 관계를 저는 ManyToMany 로 해석했고
ManyToMany 관계를 표현할 때`@JoinTable` 을 사용하기 보다는
두 개의 OneToMany, 두 개의 ManyToOne 관계를 사용해서 표현하는걸 더 합리적이라고 생각해서
LineStation 엔티티도 만들었습니다.

피드백 기다리고 있겠습니다.

Commit:
test:refactor: 인수 테스트 가독성 증대하는 방향으로 리팩터링

인수테스트는 개발자들만 보는 것이 아니라
기획자, PM 등 비개발자 이해관계자들도 함께 볼 수 있도록 만드는 것이 좋다

Commit:
feat: 어떤 머신에서든 H2 데이터베이스 사용하도록 수정

Commit:
feat: DirtiesContext 적용하기, public 또는 private 명확하게 표기하기

[1]
DirtiesContext 를 BEFORE_EACH_TEST_METHOD 로 지정하면
테스트 케이스 사이 사이에 ApplicationContext 를 초기화시켜준다

오류 원인:
WebEnvironment 에서 사용하는 LocalServerPort 는 테스트 케이스마다 바뀌지만
RestAssured.port 는 거기에 맞춰서 변하지 않아서 오류가 발생했다

[2]
public, private modifier 명시하기

Commit:
feat: 지하철 노선(Line) 도메인 엔티티 설계

Commit:
feat: LineController 만들기

Commit:
feat: LineService 구현하기

Commit:
feat: 인수 테스트 코드 작성중 1

지하철 노선 생성하기 성공하는 테스트 케이스

response 형태가 올바르지 않은 이슈 해결해야함

Commit:
feat: LineServier.findAllLine() 메서드 Stream API 사용해서 가독성 증대

Commit:
feat: ResponseEntity 로 응답하는 모든 필드에 대해서 getter 가 있어야 한다

getter 가 없는 프로퍼티는 응답 json 에서 생략된다

Commit:
feat: 지하철 노선(Line) 관련 인수 테스트 작성 2

[1]
StationController 에 upsert 기능 추가

[2]
LineController 에서 PathVariable 이름은 일치해야 한다

[3]
Response DTO 에는 no args constructor 가 있어야 파싱이 가능해진다

[4]
Line 과 Station 사이의 관계 엔티티 LineStation 프로퍼티에
`@OnDelete` 적용하기

[5]
서비스 로직에서 필요한 부분 수정

Commit:
feat: StationResponse.of() 대신에 StationResponse 생성자 사용하기

Commit:
refactor: Optional<T> 있는 부분 .orElseThrow 사용해서 개선

Commit:
feat: StationController 클래스에 Response Content-Type 추가하기

Commit:
feat: `@RequestMapping` 사용해서 경로 간소화

Commit:
feat: Controller 에 ExceptionHandler 추가 | refactor:  Service 구현 리팩터링

[1]
Controller 에 ExceptionHandler 를 추가하거나
Exception 리스트를 업데이트

[2]
Service 구현 로직을 Optional<T> 클래스에 있는 메서드를 이용해서
보다 간결하게 리팩터링

Commit:
feat: LineResponse 클래스 equals 메서드 early-return 추가

stations 개수가 다를 때 바로 return false

이후 로직을 확인할 필요가 없기도 하고
stations 길이에 맞춰서 반복문 실행하는 부분에서 발생할 수 있는
오류를 방지할 수 있다

Commit:
refactor: 불필요한 변수 제거하고 inline 으로 변경

wheejuni commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,70 @@
+        Station downStation = stationRepository.findById(lineCreateRequest.getDownStationId())
+                .orElseThrow(IllegalArgumentException::new);
+        Line line = lineRepository.save(
+                new Line(null, lineCreateRequest.getName(), lineCreateRequest.getColor(),
```
`LineCreateRequest` -> `Line` 으로 갈 수 있는 스태틱 메소드가 `Line` 쪽에 하나 있으면 좋을 것 같군요.

OptimistLabyrinth replied:
수정 완료했습니다.

wheejuni commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,70 @@
+        Line line = lineRepository.save(
+                new Line(null, lineCreateRequest.getName(), lineCreateRequest.getColor(),
+                        lineCreateRequest.getDistance()));
+        LineStation lineStationUp = lineStationRepository.save(new LineStation(upStation, line));
```
`ManyToMany` 대신 `LineStation` 으로 풀어나가신 것 아주 좋습니다. 💯

OptimistLabyrinth replied:
ㅎㅎ 감사해요

wheejuni commented on src/main/java/nextstep/subway/application/StationService.java:
```diff
@@ -21,20 +21,27 @@ public StationService(StationRepository stationRepository) {
+    public StationResponse upsert(StationRequest stationRequest) {
+        return stationRepository.findByName(stationRequest.getName())
+                .map(StationResponse::new)
+                .orElseGet(() -> new StationResponse(stationRepository.save(new Station(stationRequest.getName()))));
```
여기선 `.toStation()` 을 사용하지 않으신 이유가 있나요?

OptimistLabyrinth replied:
정신없이 하다보니 깜빡 잊었어요 ㅋㅋㅋㅋ 🤣😭

wheejuni commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,94 @@
+    public LineResponse() {
+    }
+
+    public LineResponse(Line line) {
```
생성자로 들어가기엔 너무 복잡한 로직을 담고 있습니다.
별도 스태틱 메소드로 추출해보시죠

OptimistLabyrinth replied:
수정 완료했습니다.

wheejuni reviewed (changes requested):
수고 많으셨습니다. 💯 
간단한 의견 몇 개 남깁니다. 확인 부탁드려요.

Commit:
refactor: DTO <-> Entity 변환 로직 리팩터링

Request DTO -> Domain Entity: toEntity() 메서드 사용
Domain Entity -> Response DTO: of() 메서드 사용

OptimistLabyrinth commented:
static 메서드를 많이 사용하게 되면
단위 테스트 작성 시 테스트 더블을 작성하는게 힘들어져서 안 좋지 않나요??
그다지 이상 없나요??

static 메서드의 장점은 컴파일 타임에 모든 연산을 완료해서 런타임 부하를 최소화할 수 있다고 알고 있는데요.
여기서 구현하는 of( ) 메서드는 로직상 그런 장점을 많이 발휘하기 힘들다고 생각이 들어요.
제가 생각하는게 맞나요??

이럼에도 불구하고 static 메서드를 추천해주시는 데는 뭔가 static 만의 장점이 있을 것 같아요.
조금 더 추가적으로 설명해주실 수 있나요??

wheejuni reviewed (approved):
_생성자는 값의 할당에 집중하자_ 가 제가 생각하는 기본 원칙이고 그에 기반해서 `.of()` 를 추천 드렸습니다.
연산 적으로 아껴지는 부분은 별로 고려하지 않았고요, 자바 진영이 다같이 겪고 있는 문제인데 생성자에서 자꾸 값의 할당이 아닌 다른 관심사가 생각되는 것이 개인적으로 바람직해 보이지 않습니다.

그리고 이것을 사용한다고 해서 테스트 더블이 늘어난다는 관점은 조금 동의하기 어려운데 이 부분은 제가 좀 질문을 드리고 싶기는 합니다. 어떤 면에서 그렇게 생각하셨을까요?

OptimistLabyrinth commented:
생성자는 최대한 짧게 유지하는 것이 기본적인 컨벤션이군요.
자바 뿐만 아니라 모든 언어에서 다 똑같이 적용되는 컨벤션이죠??
이건 잘 몰랐는데 새롭게 배웠어요.
감사합니다 😄🙌

두 번째 질문에 답변 드릴께요.

저도 아직 테스트 코드 작성법에 대해서 많이 아는 것은 아니지만
테스트를 작성하기 편한 구조는 DI, IoC 방식을 채택하고 있어서
테스트 코드 내에서 얼마든지 편하게 모킹한 객체로 실제 로직을 대체할 수 있어야 한다고 생각하거든요.
이런 측면에서 생각하면 싱글턴 객체, static 객체 또는 static 메서드를 
DI 방식이 아니라 직접 import 해서 사용하도록 프로덕션 코드를 구현한 경우에
그 부분의 로직을 손대지 않고 모킹해서 테스트 코드를 작성하는 게 상당히 어렵다고 느꼈거든요.

그래서 저는 일단 static 을 안 쓸 수 있으면 최대한 사용하지 않고 런타임 메서드로 사용하는게
더 바람직하다고 저 나름의 생각을 정리하게 되었습니다.

휘준님은 어떻게 생각하시나요??
